src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    internal static partial class ClrThreadPool
+    {
+        private static WaitThreadNode s_waitThreadsHead = new WaitThreadNode
+        {
+            Thread = new WaitThread()
+        };
+
+        internal static void RegisterWaitHandle(RegisteredWaitHandle handle)
+        {
+            WaitThreadNode prev;
+            WaitThreadNode current = s_waitThreadsHead;
+            do
+            {
+                if(current.Thread.RegisterWaitHandle(handle))
+                {
+                    return;
+                }
+                prev = current;
+                current = current.Next;
+            } while (current != null);
+
+            prev.Next = new WaitThreadNode
+            {
+                Thread = new WaitThread()
+            };
+            prev.Next.Thread.RegisterWaitHandle(handle);
```

Comment:
There is a race here. You have already made the newly created WaitThreadNode globally visible (prev.Next = new WaitThreadNode) so this call to RegisterWaitHandle can fail (i.e. return false)